Allow text strings to special type to date [ci drivers]

dpsutton opened the pull request:
Fixes #1576 

Ability to have text columns used as dates

Caveats:
- most databases blow up if the text string is not coercible. Some just return a null (mysql and variants).
- not all dbs are supported. All databases that go through `metabase.driver.sql-jdbc` has a default and all codepaths through `metabase.driver.sql.query-processor` has this called. Some databases provide their own query processor (bigquery, googleanalytics, druid, mongo) and these do not support this. These largely tracks the support for `:type/UNIXTimestampMilliseconds` and seconds. Mongo independently implements unix timestamps in the query processor but presumably the others are completely silent on this feature. (The unix timestamps feature is implemented in the same sql.query-processor as the feature under review here and unix timestamps are implemented independently in the mongo query processor).
- There is no feedback mechanism for when this feature is not supported. Similar to `:type/UNIXTimestampMilleseconds` on unsupported databases, not every db supports this feature. At the moment there is no mechanism to provide feedback for this. Its possible that the list of special_types is filtered to remove this, or perhaps some validation when selecting, or perhaps make the query processor aware of supported special_types and throw a proper error message. As it stands now, it will be silently ignored on databases that do not support it.
- Some databases are only partially supported. Oracle (as far as I can tell) has no time datatype. This means that the date and datetime variants are supported but it lacks a way to express `:type/ISOTimeString`. Attempting to use it will throw an invalid cast in the query processor. It would be easy to add 
```clojure
(defmethod sql.qp/cast-temporal-string [:oracle :type/ISO8601TimeString]
  [_driver _special_type expr]
  (throw (Exception. (tru "Oracle does not support ISO8601TimeString"))))
```
and perhaps I should.
- the fields aren't queryable as dates on oracle and sparksql. I haven't investigated this in depth yet but was a bit surprised. Oracle throws a NPE in the driver and sparksql throws a parse exception `DataType time() is not supported.(line 2, pos 178)`.
```clojure
(testing "are queryable as dates"
  (testing "a datetime field"
    ;; TODO: why does this fail on oracle? gives a NPE
    (mt/test-drivers (disj (sql-jdbc.tu/sql-jdbc-drivers) :oracle :sparksql)
      (is (= 1
             (count (mt/rows (mt/dataset string-times
                               (mt/run-mbql-query times
                                 {:filter   [:= [:datetime-field $ts :day] "2008-10-19"]}))))))))
  (testing "a date field"
    (mt/test-drivers (disj (sql-jdbc.tu/sql-jdbc-drivers) :oracle :sparksql)
      (is (= 1
             (count (mt/rows (mt/dataset string-times
                               (mt/run-mbql-query times
                                 {:filter   [:= [:datetime-field $d :day] "2008-10-19"]})))))))))
```


# For reviewers:

Changes should be relatively straight forward. Things I'm particularly interested in feedback on:
- should the defmethods and default implementations be in better places to give more database drivers an opportunity to hook into them? Ie, `cast-temporal-string` being in `metabase.driver.sql.query-processor` and the defaults being in `metabase.driver.sql-jdbc`.
- are any of the above caveats necessary for the first release? Do any merit follow up tickets now?

### Things to note
For future development we can use `isa?` in match expressions.
```clojure
(defn cast-field-if-needed
  "Wrap a `field-identifier` in appropriate HoneySQL expressions if it refers to a UNIX timestamp Field."
  [driver field field-identifier]
  (match [(:base_type field) (:special_type field)]
    [_ (:isa? :type/UNIXTimestampSeconds)]      (unix-timestamp->honeysql driver :seconds      field-identifier)
    [_ (:isa? :type/UNIXTimestampMilliseconds)] (unix-timestamp->honeysql driver :milliseconds field-identifier)
    [:type/Text (:isa? :type/TemporalString)]   (cast-temporal-string driver (:special_type field) field-identifier)
    :else field-identifier))
```

Very simple implementation:
```clojure
(defmethod match/emit-pattern-for-syntax [:isa? :default]
  [[_ parent]] {::match/tag ::isa? :parent parent})

(defmethod match/to-source ::isa?
  [{parent :parent} ocr]
  `(isa? ~ocr ~parent))
```

In the future, it would be simple to add the ability to add a hierarchy into this match. If its desired to add a way to capture the type, this would require moving to a custom deftype and implementing the `ISpecializeMatrix` protocol which gets a bit more involved with the matching algorithm. Not super difficult but there's not a ton of examples out there and requires knowledge of how the pattern matrices are made.

dpsutton committed:
Allow text strings to special type to date [ci drivers]

sbelak commented:
Since presumably this will only come into effect if the user manually overwrites the base type to date, I wouldn't worry about formats and assume uniformity. If either of those are not true they have bigger problems than us not picking up their dates.

camsaul commented:
> Since presumably this will only come into effect if the user manually overwrites the base type to date

@sbelak If I understand this PR correctly, they'd be leaving the base type as `:type/Text` and changing the special type to `:type/TextDate`. The base type isn't meant to be user-overwritable

camsaul commented on src/metabase/types.clj:
```diff
@@ -109,6 +109,7 @@
 (derive :type/UNIXTimestamp :type/Integer)
 (derive :type/UNIXTimestampSeconds :type/UNIXTimestamp)
 (derive :type/UNIXTimestampMilliseconds :type/UNIXTimestamp)
+(derive :type/TextDate :type/DateTime)
```
a Date is not a Date**Time**

camsaul commented on frontend/src/metabase/lib/core.js:
```diff
@@ -223,6 +223,11 @@ export const field_special_types = [
     section: t`Date and Time`,
   },
+  {
+    id: TYPE.TextDate,
+    name: t`Text as a timestamp`,
+    section: t`Date and Time`,
```
If it's a date and time then don't call it `TextDate`, call it `TextDateTime`. A `TextDate` in my mind would be `2020-11-23`

camsaul commented:
@dpsutton I don't think just having `:type/TextDate` is going to end up being sufficient. You probably want to do something like have `:type/Text` -> `:type/TemporalString` -> `:type/DateString`/`:type/DateTimeString`/`:type/TimeString` at the very least. I can tell you with 100% certainty that there are some databases out there that are not going to let you use the same function to parse `2020-11-23`, `2020-11-23T17:49:00`, and `17:49:00`.

If you just want to support `:type/TextDateTime` or `:type/DateTimeString` or whatever and add the other two later, at least leave the option to add support for the others

camsaul commented:
@dpsutton as you mentioned, we probably need different types for different formats as well. Similar to how we have both `UNIXTimestampSeconds` and `UNIXTimestampMilliseconds` -- basically the same idea but two different formats. Is the date it ISO-8601? A standard SQL timestamp like `2004-10-19 10:23:54`? `YYYYMMDD`? For a lot of DBs when parsing temporal strings we need to give it a format string.

It would be nice to be able to let users specify arbitrary format strings but that would be a pretty big change... in the mean time maybe it's better just to offer a few common formats

camsaul commented:
Related issue: https://github.com/metabase/metabase/issues/1576

camsaul commented:
Besides the things I noted, this functionality is long overdue and I'm a big 👍 on getting this PR merged. I'm happy if we just shipped ISO-8601 DateTime support for now, but I just want to make sure we're not going to end up with a dozen "my `MM/DD/YYYY`-formatted column doesn't work" issues after we ship this feature -- we should be explicit about what we're going to support and leave the door open to support more stuff in the future.

mysticaltech commented:
@camsaul Makes sense! ISO-8601 would be a great start. @dpsutton keep up the good work!

dpsutton committed:
Allow for timestamps/time/date for iso8601 strings [ci drivers]

dpsutton committed:
First pass at tests for :type/ISO8601DateTimeString [ci drivers]

dpsutton committed:
Only run ISO8601datetimestring tests on sql dbs

dpsutton committed:
Make the default a timestamp and mysql the exception [ci drivers]

dpsutton committed:
Fixes for sqlite and sqlserver

i'm not happy with sqlite yet. they are returning strings not
java.dates. But this is the same with a field who's base type is a
datetime. However, the UI allows for ordering on the date times but
not on the :type/ISO8601 fields and i'm not sure why

dpsutton committed:
Handle oracle [ci drivers]

oracle doesn't have a time type so use a similar edn dataset that
doesn't have the time column

dpsutton committed:
Move dataset definitions inline [ci drivers]

dpsutton committed:
correct oracle datetime format [ci drivers]

sparksql seems to have issues with time so just punt on that for now
by running in the test that doesn't have the time column

dpsutton committed:
wack-a-mole [ci drivers]

dpsutton committed:
Fix tests for redshift, oracle, and sparksql [ci drivers]

dpsutton committed:
Oracle is funky [ci drivers]

dpsutton committed:
need to investigate what's going on with oracle here [ci drivers]

text fields with special_type :type/ISO8601DateTimeString are queryable as dates a datetime field
:oracle
expected: (= 1 (count (mt/rows (mt/dataset string-times (mt/run-mbql-query times {:filter [:= [:datetime-field $ts :day] "2008-10-19"]})))))
  actual: java.lang.NullPointerException: null
 at oracle.jdbc.driver.T4CTTIdcb.fillupAccessors (T4CTTIdcb.java:1041)
    oracle.jdbc.driver.T4CTTIdcb.receiveCommon (T4CTTIdcb.java:221)
    oracle.jdbc.driver.T4CTTIdcb.receive (T4CTTIdcb.java:161)
    oracle.jdbc.driver.T4C8Oall.readDCB (T4C8Oall.java:991)
    oracle.jdbc.driver.T4CTTIfun.receive (T4CTTIfun.java:560)
    oracle.jdbc.driver.T4CTTIfun.doRPC (T4CTTIfun.java:252)
    oracle.jdbc.driver.T4C8Oall.doOALL (T4C8Oall.java:612)
    oracle.jdbc.driver.T4CPreparedStatement.doOall8 (T4CPreparedStatement.java:226)
    oracle.jdbc.driver.T4CPreparedStatement.doOall8 (T4CPreparedStatement.java:59)
    oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe (T4CPreparedStatement.java:747)
    oracle.jdbc.driver.OracleStatement.executeMaybeDescribe (OracleStatement.java:904)
    oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout (OracleStatement.java:1082)
    oracle.jdbc.driver.OraclePreparedStatement.executeInternal (OraclePreparedStatement.java:3780)
    oracle.jdbc.driver.T4CPreparedStatement.executeInternal (T4CPreparedStatement.java:1343)
    oracle.jdbc.driver.OraclePreparedStatement.executeQuery (OraclePreparedStatement.java:3822)
    oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery (OraclePreparedStatementWrapper.java:1165)
    com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery (NewProxyPreparedStatement.java:431)
    metabase.driver.sql_jdbc.execute$eval71770$fn__71771.invoke (execute.clj:267)
    clojure.lang.MultiFn.invoke (MultiFn.java:234)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic (execute.clj:392)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke (execute.clj:377)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic (execute.clj:386)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke (execute.clj:377)
    metabase.driver.sql_jdbc$eval102870$fn__102871.invoke (sql_jdbc.clj:49)
    metabase.driver.oracle$eval187218$fn__187219.invoke (oracle.clj:268)
    clojure.lang.MultiFn.invoke (MultiFn.java:244)
    metabase.query_processor.context$executef.invokeStatic (context.clj:59)
    metabase.query_processor.context$executef.invoke (context.clj:48)
    metabase.query_processor.context.default$default_runf.invokeStatic (default.clj:69)
    metabase.query_processor.context.default$default_runf.invoke (default.clj:67)
    metabase.query_processor.context$runf.invokeStatic (context.clj:45)
    metabase.query_processor.context$runf.invoke (context.clj:39)
    metabase.query_processor.reducible$pivot.invokeStatic (reducible.clj:34)
    metabase.query_processor.reducible$pivot.invoke (reducible.clj:31)

dpsutton committed:
sparksql has some issues querying on datetime [ci drivers]

text fields with special_type :type/ISO8601DateTimeString are queryable as dates a datetime field
:sparksql: org.apache.spark.sql.catalyst.parser.ParseException:
DataType time() is not supported.(line 2, pos 178)

== SQL ==
-- Metabase
SELECT `t1`.`id` AS `id`, `t1`.`name` AS `name`, CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) AS `ts`, CAST(CAST(`t1`.`d` AS date) AS timestamp) AS `d`, CAST(CAST(`t1`.`t` AS time) AS timestamp) AS `t` FROM `string_times`.`times` `t1` WHERE (CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) >= CAST(to_utc_timestamp('2008-10-19 00:00:00', 'UTC') AS timestamp) AND CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) < CAST(to_utc_timestamp('2008-10-20 00:00:00', 'UTC') AS timestamp)) LIMIT 1048576
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------^^^

expected: (= 1 (count (mt/rows (mt/dataset string-times (mt/run-mbql-query times {:filter [:= [:datetime-field $ts :day] "2008-10-19"]})))))
  actual: java.sql.SQLException: org.apache.spark.sql.catalyst.parser.ParseException:
DataType time() is not supported.(line 2, pos 178)

== SQL ==
-- Metabase
SELECT `t1`.`id` AS `id`, `t1`.`name` AS `name`, CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) AS `ts`, CAST(CAST(`t1`.`d` AS date) AS timestamp) AS `d`, CAST(CAST(`t1`.`t` AS time) AS timestamp) AS `t` FROM `string_times`.`times` `t1` WHERE (CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) >= CAST(to_utc_timestamp('2008-10-19 00:00:00', 'UTC') AS timestamp) AND CAST(CAST(`t1`.`ts` AS timestamp) AS timestamp) < CAST(to_utc_timestamp('2008-10-20 00:00:00', 'UTC') AS timestamp)) LIMIT 1048576
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------^^^

 at org.apache.hive.jdbc.HiveStatement.execute (HiveStatement.java:297)
    org.apache.hive.jdbc.HiveStatement.executeQuery (HiveStatement.java:392)
    org.apache.hive.jdbc.HivePreparedStatement.executeQuery (HivePreparedStatement.java:109)
    com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery (NewProxyPreparedStatement.java:431)
    metabase.driver.sql_jdbc.execute$eval71770$fn__71771.invoke (execute.clj:267)
    clojure.lang.MultiFn.invoke (MultiFn.java:234)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic (execute.clj:392)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke (execute.clj:377)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invokeStatic (execute.clj:386)
    metabase.driver.sql_jdbc.execute$execute_reducible_query.invoke (execute.clj:377)
    metabase.driver.sql_jdbc$eval102870$fn__102871.invoke (sql_jdbc.clj:49)
    metabase.driver.sparksql$eval187174$fn__187176.invoke (sparksql.clj:142)
    clojure.lang.MultiFn.invoke (MultiFn.java:244)
    metabase.query_processor.context$executef.invokeStatic (context.clj:59)
    metabase.query_processor.context$executef.invoke (context.clj:48)
    metabase.query_processor.context.default$default_runf.invokeStatic (default.clj:69)
    metabase.query_processor.context.default$default_runf.invoke (default.clj:67)
    metabase.query_processor.context$runf.invokeStatic (context.clj:45)
    metabase.query_processor.context$runf.invoke (context.clj:39)
    metabase.query_processor.reducible$pivot.invokeStatic (reducible.clj:34)
    metabase.query_processor.reducible$pivot.invoke (reducible.clj:31)
    metabase.query_processor.middleware.mbql_to_native$mbql__GT_native$fn__57878.invoke (mbql_to_native.clj:26)
    metabase.query_processor.middleware.check_features$check_features$fn__57065.invoke (check_features.clj:42)
    metabase.query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__58077.invoke (optimize_datetime_filters.clj:133)
    metabase.query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__55756.invoke (auto_parse_filter_values.clj:44)
    metabase.query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__59813.invoke (wrap_value_literals.clj:149)
    metabase.query_processor.middleware.annotate$add_column_info$fn__55502.invoke (annotate.clj:575)
    metabase.query_processor.middleware.permissions$check_query_permissions$fn__56924.invoke (permissions.clj:70)
    metabase.query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__58633.invoke (pre_alias_aggregations.clj:40)
    metabase.query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__57154.invoke (cumulative_aggregations.clj:61)
    metabase_enterprise.sandbox.query_processor.middleware.row_level_restrictions$apply_row_level_permissions$fn__60343.invoke (row_level_restrictions.clj:292)
    metabase.query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__58912.invoke (resolve_joined_fields.clj:36)
    metabase.query_processor.middleware.resolve_joins$resolve_joins$fn__59239.invoke (resolve_joins.clj:183)
    metabase.query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__50719.invoke (add_implicit_joins.clj:254)
    metabase.query_processor.middleware.large_int_id$convert_id_to_string$fn__57823.invoke (large_int_id.clj:44)
    metabase.query_processor.middleware.limit$limit$fn__57856.invoke (limit.clj:38)
    metabase.query_processor.middleware.format_rows$format_rows$fn__57795.invoke (format_rows.clj:84)
    metabase.query_processor.middleware.desugar$desugar$fn__57226.invoke (desugar.clj:22)
    metabase.query_processor.middleware.binning$update_binning_strategy$fn__56123.invoke (binning.clj:229)
    metabase.query_processor.middleware.resolve_fields$resolve_fields$fn__56714.invoke (resolve_fields.clj:24)
    metabase.query_processor.middleware.add_dimension_projections$add_remapping$fn__50239.invoke (add_dimension_projections.clj:318)
    metabase.query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__50462.invoke (add_implicit_clauses.clj:141)
    metabase_enterprise.sandbox.query_processor.middleware.row_level_restrictions$apply_row_level_permissions$fn__60343.invoke (row_level_restrictions.clj:292)
    metabase.query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__50882.invoke (add_source_metadata.clj:105)
    metabase_enterprise.sandbox.query_processor.middleware.column_level_perms_check$maybe_apply_column_level_perms_check$fn__60025.invoke (column_level_perms_check.clj:24)
    metabase.query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__58858.invoke (reconcile_breakout_and_order_by_bucketing.clj:98)

dpsutton committed:
Remove sparksql can't query on date [ci drivers]

mysticaltech commented:
Keep up the good work @dpsutton! Presto support would be mind-bending.

dpsutton commented:
@mysticaltech all sql databases should get this by default. I haven't run presto locally with this change but the `alternative-date-tests` run against the presto driver:

```
# output from the presto driver tests
Testing ********.query-processor-test.alternative-date-test
Sync :presto Database sad-toucan-incidents took 10.6 s
Sync :presto Database tupac-sightings took 17.2 s
load-data for h2 string-times times took 5.7 ms # <- this is the db with the string date types
Sync :h2 Database string-times took 266.6 ms
```

It's running against the `string-times` sample database so it seems this is working for presto.

sbelak commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -209,6 +210,16 @@
   (fn [driver seconds-or-milliseconds _] [(driver/dispatch-on-initialized-driver driver) seconds-or-milliseconds])
   :hierarchy #'driver/hierarchy)
 
+(defmulti cast-temporal-string
```
I think it's reasonable to have these here. Non-SQL as they are in a lot of ways an artefact of SQL semantics. If we ever come to a situation that we want to support this for other drivers, we can always move it to the main driver namespace, but for now I think smaller surface area wins.

dpsutton replied:
thanks @sbelak . It felt right and kinda mimiced what the unix timestamp was doing but was worried i was foreclosing other databases unnecessarily.

mysticaltech replied:
Well done! 🏆

sbelak reviewed (approved):
Concur with Cam. Looks good and a great addition to MB!

camsaul reviewed (approved):
Awesome
